File: agent/cursor/src/session_start.ts
import { spawn } from "node:child_process";
import type { AgentMessage } from "@valon-technologies/gestalt";

export const SESSION_START_PREFIX = "__gestalt.lifecycle.sessionStart";
export const SESSION_START_ADDITIONAL_CONTEXT_KEY = `${SESSION_START_PREFIX}.additionalContext`;
const RESERVED_METADATA_KEYS = new Set(["cwd", "workspacePath", "worktreePath"]);

const DEFAULT_ENV_KEYS = [
  "HOME",
  "PATH",
  "SHELL",
  "TMPDIR",
  "USER",
  "LOGNAME",
  "LANG",
  "LC_ALL",
] as const;

export async function runSessionStartHooks(
  sessionStart: unknown,
  metadata: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  const hooks = sessionStartHooks(sessionStart);
  if (hooks.length === 0) {
    return { ...metadata };
  }
  const merged: Record<string, unknown> = { ...metadata };
  const contextChunks: string[] = [];
  for (const hook of hooks) {
    const { result, additionalContext } = await runHook(hook);
    merged[`${SESSION_START_PREFIX}.results.${hookId(hook)}`] = result;
    if (additionalContext) {
      contextChunks.push(additionalContext);
    }
  }
  if (contextChunks.length > 0) {
    merged[SESSION_START_ADDITIONAL_CONTEXT_KEY] = contextChunks.join("\n\n");
  }
  return merged;
}

export function validateSessionStartUserMetadata(
  metadata: Record<string, unknown> | undefined,
): void {
  if (metadata === undefined) {
    return;
  }
  for (const key of Object.keys(metadata)) {
    if (key.startsWith(SESSION_START_PREFIX) || RESERVED_METADATA_KEYS.has(key)) {
      throw new Error(
        `agent session metadata key ${JSON.stringify(key)} is reserved for Gestalt lifecycle or workspace data`,
      );
    }
  }
}

export function prependSessionStartContext(
  messages: readonly AgentMessage[],
  metadata: Record<string, unknown>,
): AgentMessage[] {
  const context = String(metadata[SESSION_START_ADDITIONAL_CONTEXT_KEY] ?? "").trim();
  if (!context) {
    return [...messages];
  }
  return [
    {
      role: "system",
      text: `Session start context:\n\n${context}`,
      metadata: { source: SESSION_START_PREFIX },
    },
    ...messages,
  ];
}

function sessionStartHooks(sessionStart: unknown): Record<string, unknown>[] {
  if (!sessionStart || typeof sessionStart !== "object") {
    return [];
  }
  const hooks = (sessionStart as { hooks?: unknown }).hooks;
  return Array.isArray(hooks) ? hooks.filter(isRecord) : [];
}

async function runHook(hook: Record<string, unknown>): Promise<{
  result: Record<string, unknown>;
  additionalContext: string;
}> {
  const id = hookId(hook);
  const type = String(hook.type ?? "command").trim() || "command";
  if (type !== "command") {
    throw new Error(`sessionStart hook ${JSON.stringify(id)} type ${JSON.stringify(type)} is not supported`);
  }
  const command = Array.isArray(hook.command)
    ? hook.command.map((part) => String(part)).filter((part) => part.trim())
    : [];
  if (command.length === 0) {
    throw new Error(`sessionStart hook ${JSON.stringify(id)} command is required`);
  }
  const timeout = String(hook.timeout ?? "");
  const completed = await runCommand({
    command: command[0]!,
    args: command.slice(1),
    cwd: String(hook.cwd ?? "").trim() || undefined,
    env: hookEnv(isRecord(hook.env) ? hook.env : {}),
    timeoutMs: parseTimeoutMs(timeout),
  });
  if (completed.code !== 0) {
    const detail = completed.stderr.trim() || completed.stdout.trim() || `exit code ${completed.code}`;
    throw new Error(`sessionStart hook ${JSON.stringify(id)} failed: ${detail}`);
  }
  const output = isRecord(hook.output) ? hook.output : {};
  const stdoutPayload = jsonStdoutPayload(completed.stdout);
  const result: Record<string, unknown> = {
    status: "succeeded",
    exitCode: completed.code,
    timeout,
    timedOut: false,
  };
  if (output.metadata === true) {
    if (isRecord(stdoutPayload.metadata)) {
      result.metadata = stdoutPayload.metadata;
    }
    result.stdout = completed.stdout;
    result.stderr = completed.stderr;
  }
  const payloadContext = stdoutPayload.additionalContext;
  return {
    result,
    additionalContext:
      output.additionalContext === true
        ? payloadContext === undefined
          ? completed.stdout.trim()
          : String(payloadContext).trim()
        : "",
  };
}

function runCommand(input: {
  command: string;
  args: string[];
  cwd: string | undefined;
  env: NodeJS.ProcessEnv;
  timeoutMs: number | undefined;
}): Promise<{ code: number; stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    const child = spawn(input.command, input.args, {
      cwd: input.cwd,
      env: input.env,
      stdio: ["ignore", "pipe", "pipe"],
    });
    let stdout = "";
    let stderr = "";
    let timeout: ReturnType<typeof setTimeout> | undefined;
    if (input.timeoutMs !== undefined) {
      timeout = setTimeout(() => {
        child.kill("SIGTERM");
        reject(new Error(`sessionStart hook timed out after ${input.timeoutMs}ms`));
      }, input.timeoutMs);
    }
    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (chunk) => {
      stdout += chunk;
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk;
    });
    child.on("error", reject);
    child.on("close", (code) => {
      if (timeout) {
        clearTimeout(timeout);
      }
      resolve({ code: code ?? 1, stdout, stderr });
    });
  });
}

function hookEnv(explicit: Record<string, unknown>): NodeJS.ProcessEnv {
  const env: NodeJS.ProcessEnv = {};
  for (const key of DEFAULT_ENV_KEYS) {
    if (process.env[key] !== undefined) {
      env[key] = process.env[key];
    }
  }
  for (const [key, value] of Object.entries(explicit)) {
    env[key] = String(value);
  }
  return env;
}

function parseTimeoutMs(value: string): number | undefined {
  const trimmed = value.trim();
  if (!trimmed) {
    return undefined;
  }
  if (trimmed.endsWith("ms")) {
    return Number(trimmed.slice(0, -2));
  }
  if (trimmed.endsWith("s")) {
    return Number(trimmed.slice(0, -1)) * 1000;
  }
  if (trimmed.endsWith("m")) {
    return Number(trimmed.slice(0, -1)) * 60_000;
  }
  return Number(trimmed) * 1000;
}

function hookId(hook: Record<string, unknown>): string {
  return String(hook.id ?? "").trim();
}

function jsonStdoutPayload(stdout: string): Record<string, unknown> {
  const trimmed = stdout.trim();
  if (!trimmed) {
    return {};
  }
  try {
    const payload: unknown = JSON.parse(trimmed);
    return isRecord(payload) ? payload : {};
  } catch {
    return {};
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: agent/claude/internals/session_start.py
from __future__ import annotations

import json
import os
import subprocess
from collections.abc import Mapping
from typing import Any

import gestalt

RESERVED_PREFIX = "__gestalt.lifecycle.sessionStart"
RESULTS_PREFIX = f"{RESERVED_PREFIX}.results"
ADDITIONAL_CONTEXT_KEY = f"{RESERVED_PREFIX}.additionalContext"
RESERVED_METADATA_KEYS = {"cwd", "workspacePath", "worktreePath"}

_DEFAULT_ENV_KEYS = ("HOME", "PATH", "SHELL", "TMPDIR", "USER", "LOGNAME", "LANG", "LC_ALL")


def run_session_start_hooks(
    session_start: gestalt.AgentSessionStartConfig, metadata: dict[str, Any]
) -> dict[str, Any]:
    hooks = list(session_start.hooks)
    if not hooks:
        return dict(metadata)
    merged = dict(metadata)
    context_chunks: list[str] = []
    for hook in hooks:
        result, additional_context = _run_hook(hook)
        hook_id = hook.id.strip()
        merged[f"{RESULTS_PREFIX}.{hook_id}"] = result
        if additional_context:
            context_chunks.append(additional_context)
    if context_chunks:
        merged[ADDITIONAL_CONTEXT_KEY] = "\n\n".join(context_chunks)
    return merged


def validate_session_start_user_metadata(metadata: dict[str, Any] | None) -> None:
    if metadata is None:
        return
    for key in metadata:
        key = str(key)
        if key.startswith(RESERVED_PREFIX) or key in RESERVED_METADATA_KEYS:
            raise ValueError(f"agent session metadata key {key!r} is reserved for Gestalt lifecycle or workspace data")


def prepend_session_start_context(messages: list[dict[str, Any]], metadata: dict[str, Any]) -> list[dict[str, Any]]:
    context = str(metadata.get(ADDITIONAL_CONTEXT_KEY) or "").strip()
    if not context:
        return messages
    return [
        {"role": "system", "text": f"Session start context:\n\n{context}", "metadata": {"source": RESERVED_PREFIX}},
        *messages,
    ]


def _run_hook(hook: gestalt.AgentSessionStartHook) -> tuple[dict[str, Any], str]:
    hook_id = hook.id.strip()
    hook_type = hook.type.strip() or "command"
    if hook_type != "command":
        raise ValueError(f"sessionStart hook {hook_id!r} type {hook_type!r} is not supported")
    command = [str(part) for part in hook.command if str(part).strip()]
    if not command:
        raise ValueError(f"sessionStart hook {hook_id!r} command is required")
    timeout = _parse_timeout(hook.timeout)
    try:
        completed = subprocess.run(
            command,
            cwd=hook.cwd or None,
            env=_hook_env(hook.env),
            text=True,
            capture_output=True,
            timeout=timeout,
            check=False,
        )
    except subprocess.TimeoutExpired as exc:
        timeout_detail = str(hook.timeout or timeout or "").strip()
        raise RuntimeError(f"sessionStart hook {hook_id!r} timed out after {timeout_detail}") from exc
    stdout = completed.stdout or ""
    stderr = completed.stderr or ""
    if completed.returncode != 0:
        detail = stderr.strip() or stdout.strip() or f"exit code {completed.returncode}"
        raise RuntimeError(f"sessionStart hook {hook_id!r} failed: {detail}")
    stdout_payload = _json_stdout_payload(stdout)
    result: dict[str, Any] = {
        "status": "succeeded",
        "exitCode": completed.returncode,
        "timeout": hook.timeout,
        "timedOut": False,
    }
    if hook.output is not None and hook.output.metadata:
        payload_metadata = stdout_payload.get("metadata")
        if isinstance(payload_metadata, dict):
            result["metadata"] = payload_metadata
        result["stdout"] = stdout
        result["stderr"] = stderr
    additional_context = ""
    if hook.output is not None and hook.output.additional_context:
        payload_context = stdout_payload.get("additionalContext")
        additional_context = str(payload_context).strip() if payload_context is not None else stdout.strip()
    return result, additional_context


def _json_stdout_payload(stdout: str) -> dict[str, Any]:
    stdout = stdout.strip()
    if not stdout:
        return {}
    try:
        payload = json.loads(stdout)
    except json.JSONDecodeError:
        return {}
    return payload if isinstance(payload, dict) else {}


def _hook_env(explicit: Mapping[str, str]) -> dict[str, str]:
    env = {key: value for key in _DEFAULT_ENV_KEYS if (value := os.environ.get(key)) is not None}
    env.update(explicit)
    return env


def _parse_timeout(value: str) -> float | None:
    value = value.strip()
    if not value:
        return None
    if value.endswith("ms"):
        return float(value[:-2]) / 1000
    if value.endswith("s"):
        return float(value[:-1])
    if value.endswith("m"):
        return float(value[:-1]) * 60
    return float(value)
